build: build only the necessary arch for roachtests

RaduBerinde opened the pull request:
The nightly roachtests (and other builds) use
`roachtest_compile_bits.sh` to build the necessary roachtest binaries. This script builds them for three architectures (amd64, amd64-fips, arm64), which takes almost 1 hour. Only one architecture is used for any given build.

This commit updates the script to take the necessary arch as an argument.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110949)
<!-- Reviewable:end -->

RaduBerinde commented:
> All three architectures are conditionally required for a single nightly run. The plan was to pass metamorphic-arm64-probability and metamorphic-fips-probability so that each run executes compatible tests on one of the three available architectures

Will that be the case for both GCE and AWS? In any case, when we start using those flags, we can simply update the script to add the necessary arches as arguments (depending on the flags).

I'd still like to make this change so I can use the script to build just amd64 with coverage.

renatolabs commented on build/teamcity/cockroach/ci/tests/roachtest_stress_impl.sh:
```diff
@@ -10,7 +10,11 @@ if [[ ! -f ~/.ssh/id_rsa.pub ]]; then
+arch=amd64
+if [[ ${FIPS_ENABLED:-0} == 1 ]]; then
+  arch=amd64-fips
+fi
```
AFAIK, this build is only triggered manually. Consider whether the `arch` could be passed directly as an environment variable via the TC UI (allowing us to stress tests on ARM too, for example).

renatolabs commented on build/teamcity/cockroach/ci/tests/roachtest_stress_impl.sh:
```diff
@@ -10,7 +10,11 @@ if [[ ! -f ~/.ssh/id_rsa.pub ]]; then
+if [[ ${FIPS_ENABLED:-0} == 1 ]]; then
+  arch=amd64-fips
+fi
+$root/build/teamcity/cockroach/nightlies/roachtest_compile_bits.sh
```
Missing `$arch`.

renatolabs reviewed (approved):
I'm fine with this change, especially since:

> we can simply update the script to add the necessary arches as arguments

it seems like we'll be able to add `arm64` and `fips` back to the nightly build easily when the metamorphic flags are introduced.

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110949#-:-NeoJsQ8AZzMKkSiA_L0:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @rickystewart, and @srosenberg)

___
*[`build/teamcity/cockroach/ci/tests/roachtest_stress_impl.sh` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110949#gh-1332098904:-NeoJQmBBz4AqOGll--E:bzh1gke) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7e4c299cf170fc302ad7c26bdf082f867cf8cf5/build/teamcity/cockroach/ci/tests/roachtest_stress_impl.sh#L16)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

AFAIK, this build is only triggered manually. Consider whether the `arch` could be passed directly as an environment variable via the TC UI (allowing us to stress tests on ARM too, for example). 
</blockquote></details>

It's hard to develop against things that aren't merged yet.. The best way to support ARM would depend on how we will pass that information on to roachtest, presumably with a flag. Ideally we'd use that flag instead of relying on the TC build setup to have the `arch` and the flags "in sync".
___
*[`build/teamcity/cockroach/ci/tests/roachtest_stress_impl.sh` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110949#gh-1332099087:-NeoJQZY8F2MT6ej_iL_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7e4c299cf170fc302ad7c26bdf082f867cf8cf5/build/teamcity/cockroach/ci/tests/roachtest_stress_impl.sh#L17)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Missing `$arch`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

srosenberg commented:
> Will that be the case for both GCE and AWS? 

Yep, both gce and aws support running all three. 

> In any case, when we start using those flags, we can simply update the script to add the necessary arches as arguments (depending on the flags).

Sure, but we will be back to square one. That is, a single run of roachtest nightly (gce, aws or azure) will need the binaries for all three architectures. In addition, it needs one compiled with `crdb_test`, plus one instrumented with coverage.

> I'd still like to make this change so I can use the script to build just amd64 with coverage.

Why not just add another build of amd64 with coverage? The conditional compilation won't solve the problem of needing all these binaries before `roachtest run` is executed. 

> The best way to support ARM would depend on how we will pass that information on to roachtest, presumably with a flag. Ideally we'd use that flag instead of relying on the TC build setup to have the arch and the flags "in sync".

The flag is already passed implicitly; it currently defaults to `0`. Let's say, tomorrow we change it to `--metamorphic-arm64-probability .25`. Subsequently, `roachtest run` will fail since it's unable to find the arm binary.

RaduBerinde committed:
build: build only the necessary arch for roachtests

The nightly roachtests (and other builds) use
`roachtest_compile_bits.sh` to build the necessary roachtest binaries.
This script builds them for three architectures (amd64, amd64-fips,
arm64), which takes almost 1 hour. Only one architecture is used for
any given build.

This commit updates the script to take the necessary arch as an
argument.

Epic: none
Release note: None

RaduBerinde commented:
The coverage build will only need `amd64`. I don't want it to waste 2hrs each time building the other binaries. My only options are to make this script more flexible (I plan to add support for a `-with-coverage` flag on top of this change), or copy most of it to a new script. I think the former is clearly better. And if/when we move to separate out the builds for the various platforms, the updated script will be useful.

> The flag is already passed implicitly; it currently defaults to 0. Let's say, tomorrow we change it to --metamorphic-arm64-probability .25.

The scripts currently do NOT pass `--metamorphic-arm64-probability` at all. When they start doing that, my suggestion would be (as part of that same change) to dynamically choose to add the arm build argument when that value is not 0. That way if you e.g. run a manual build with no arm, we won't build it unnecessarily.

srosenberg commented:
> The scripts currently do NOT pass `--metamorphic-arm64-probability` at all. When they start doing that, my suggestion would be (as part of that same change) to dynamically choose to add the arm build argument when that value is not 0. That way if you e.g. run a manual build with no arm, we won't build it unnecessarily.

Sounds good, I can follow up with a PR for that. Let's merge your changes as is.

RaduBerinde commented:
bors r+
